// @awsui/components-react/test-utils/dom/calendar/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var dom_1 = require("@awsui/test-utils-core/dom");
var styles_selectors_js_1 = require("../../../calendar/styles.selectors.js");
var button_1 = require("../button");
var CalendarWrapper = /** @class */ (function (_super) {
    __extends(CalendarWrapper, _super);
    function CalendarWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Returns a day container on the calendar.
     *
     * @param row 1-based row index of the day.
     * @param column 1-based column index of the day.
     */
    CalendarWrapper.prototype.findDateAt = function (row, column) {
        return this.find(".".concat(styles_selectors_js_1.default['calendar-week'], ":nth-child(").concat(row, ") .").concat(styles_selectors_js_1.default['calendar-day'], ":nth-child(").concat(column, ")"));
    };
    CalendarWrapper.prototype.findHeader = function () {
        return this.findByClassName(styles_selectors_js_1.default['calendar-header']);
    };
    CalendarWrapper.prototype.findPreviousMonthButton = function () {
        return this.findComponent(".".concat(styles_selectors_js_1.default['calendar-prev-month-btn']), button_1.default);
    };
    CalendarWrapper.prototype.findNextMonthButton = function () {
        return this.findComponent(".".concat(styles_selectors_js_1.default['calendar-next-month-btn']), button_1.default);
    };
    CalendarWrapper.prototype.findSelectedDate = function () {
        return this.find(".".concat(styles_selectors_js_1.default['calendar-day-selected']));
    };
    CalendarWrapper.rootSelector = styles_selectors_js_1.default.root;
    return CalendarWrapper;
}(dom_1.ComponentWrapper));
exports.default = CalendarWrapper;
